examples/ecs/computed_states.rs

```diff
@@ -0,0 +1,674 @@
+//! This example illustrates the use of [`ComputedStates`] for more complex state handling patterns.
+//!
+//! In this case, we'll be implementing the following pattern:
+//! - The game will start in a `Menu` state, which we can return to with `Esc`
+//! - From there, we can enter the game - where our bevy symbol moves around and changes color
+//! - While in game, we can pause and unpause the game using `Space`
+//! - We can also toggle "Turbo Mode" with the `T` key - where the movement and color changes are all faster. This
+//!   is retained between pauses, but not if we exit to the main menu.
+//!
+//! In addition, we want to enable a "tutorial" mode, which will involve it's own state that is toggled in the main menu.
+//! This will display instructions about movement and turbo mode when in game and unpaused, and instructions on how to unpause when paused.
+
+use bevy::prelude::*;
+
+// To begin, we want to define our state objects.
+#[derive(Debug, Clone, Copy, Default, Eq, PartialEq, Hash, States)]
+enum AppState {
+    #[default]
+    Menu,
+    // Unlike in the `states` example, we're adding more data in this
+    // version of our AppState. In this case, we actually have
+    // 4 distinct "InGame" states - unpaused and no turbo, paused and no
+    // turbo, unpaused and turbo and paused and turbo.
+    InGame {
+        paused: bool,
+        turbo: bool,
+    },
+}
+
+// The tutorial state object, on the other hand, is a fairly simple enum.
+#[derive(Debug, Clone, Copy, Default, Eq, PartialEq, Hash, States)]
+enum TutorialState {
+    #[default]
+    Active,
+    Inactive,
+}
+
+// Because we have 4 distinct "InGame" states, we're going to define
+// a separate "InGame" type and implement `ComputedStates` for it.
+// This allows us to only need to check against one type
+// when otherwise we'd need to check against multiple.
+#[derive(Debug, Clone, Copy, Eq, PartialEq, Hash)]
+struct InGame;
+
+impl ComputedStates for InGame {
+    // Our computed state depends on `AppState`, so we need to specify it as the SourceStates type.
+    type SourceStates = AppState;
+
+    // The compute function takes in the `SourceStates` wrapped in `Option`'s, in this case the complex type
+    // signature actually boils down to `Option<AppState>`, but I wanted to leave the complex version visible
+    // so you are aware of what it is in case it is generated for you by your IDE.
+    fn compute(
+        sources: <<Self as ComputedStates>::SourceStates as bevy_internal::ecs::schedule::StateSet>::Optionals,
+    ) -> Option<Self> {
+        // You might notice that InGame has no values - instead, in this case, the `State<InGame>` resource only exists
+        // if the `compute` function would return `Some` - so only when we are in game.
+        match sources {
+            Some(AppState::InGame { .. }) => Some(Self),
```

Comment:
```suggestion
            // No matter what the value of `paused` or `turbo` is, we're still in the game rather than a menu
            Some(AppState::InGame { .. }) => Some(Self),
```